src/art.rs

```diff
@@ -0,0 +1,913 @@
+
+                    for i in 0..inner.meta.num_children {
+                        let child = &inner.children[i];
+                        if !child.is_empty() {
+                            node48.children[node48.meta.num_children] =
+                                mem::take(&mut inner.children[i]);
+                            node48.keys[i] = (node48.meta.num_children + 1) as u8;
+                            node48.meta.num_children += 1;
+                        }
+                    }
+
+                    *self = Node::Inner(Box::new(node48));
+                }
+            },
+            Node::Leaf(_) | Node::Empty => {} // do nothing
+        }
+    }
+
+    pub fn insert(&mut self, key: &[u8], value: V, mut depth: usize) -> Option<V> {
+        match *self {
+            // @spec: Usually, the leaf can
+            //        simply be inserted into an existing inner node, after growing
+            //        it if necessary.
+            Node::Empty => {
+                let leaf = LeafNode::new(key, value);
+                *self = Node::Leaf(Box::new(leaf));
+                None
+            }
+            // @spec: If, because of lazy expansion,
+            //        an existing leaf is encountered, it is replaced by a new
+            //        inner node storing the existing and the new leaf
+            Node::Leaf(ref mut leaf) => {
+                if leaf.matches(key) {
+                    let old_value = mem::replace(&mut leaf.value, value);
+                    return Some(old_value);
+                }
+
+                // Create a new Inner Node to contain the new Leaf and the current node.
+                let new_leaf = LeafNode::new(key, value);
+
+                // Determine the longest common prefix between our current node and the key
+                let limit = leaf.longest_common_prefix(&new_leaf, depth);
+
+                let mut partial_new = [0u8; MAX_PREFIX_LEN];
+                partial_new[..min(MAX_PREFIX_LEN, limit)].copy_from_slice(&key[depth..(min(MAX_PREFIX_LEN, limit) + depth)]);
+                // for i in 0..min(MAX_PREFIX_LEN, limit) {
+                //     partial_new[i] = key[depth + i];
+                // }
+
+                let mut n4: InnerNode<V> = InnerNode::new_node4();
+                n4.meta.prefix_len = limit;
+                n4.meta.prefix = partial_new;
+                let n4_node: Node<V> = Node::Inner(Box::new(n4));
+
+                let old_node = mem::replace(self, n4_node);
+                match old_node {
+                    Node::Leaf(old_leaf) => match self {
+                        Node::Inner(inner) => {
+                            inner.add_child(old_leaf.as_ref().key[depth], Node::Leaf(old_leaf));
+                            inner.add_child(new_leaf.key[depth], Node::Leaf(Box::new(new_leaf)));
+                        }
+                        _ => unreachable!(),
+                    },
+                    _ => unreachable!(),
+                }
```

Comment:
I don't really like this pattern of replace and then rematching the node. It might be better to take ownership at the match and then put it back at the end of the method.
I did a quick and dirty refactor here: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d870926f382ec6f0bdcfd4a6da74106a
Look it over if you want to adopt it because I haven't tried to fully understand the code.

Reply:
Thanks! This is fixed in https://github.com/surrealdb/tart/pull/2